File: packages/ocp-shuffle/src/encoding.ts
import type { ElGamalCiphertext, GroupElement, Scalar } from "@onchainpoker/ocp-crypto";
import { concatBytes, groupElementFromBytes, groupElementToBytes, scalarFromBytes, scalarToBytes } from "@onchainpoker/ocp-crypto";

export function u16ToBytesLE(x: number): Uint8Array {
  if (!Number.isInteger(x) || x < 0 || x > 0xffff) throw new Error("u16ToBytesLE: out of range");
  return new Uint8Array([x & 0xff, (x >> 8) & 0xff]);
}

export function u16FromBytesLE(bytes: Uint8Array): number {
  if (!(bytes instanceof Uint8Array) || bytes.length !== 2) throw new Error("u16FromBytesLE: expected 2 bytes");
  return bytes[0]! | (bytes[1]! << 8);
}

export function encodePoint(p: GroupElement): Uint8Array {
  return groupElementToBytes(p);
}

export function decodePoint(bytes: Uint8Array): GroupElement {
  return groupElementFromBytes(bytes);
}

export function encodeScalar(s: Scalar): Uint8Array {
  return scalarToBytes(s);
}

export function decodeScalar(bytes: Uint8Array): Scalar {
  return scalarFromBytes(bytes);
}

export function encodeCiphertext(ct: ElGamalCiphertext): Uint8Array {
  const c1 = encodePoint(ct.c1);
  const c2 = encodePoint(ct.c2);
  return concatBytes(c1, c2);
}

export function decodeCiphertext(bytes: Uint8Array): ElGamalCiphertext {
  if (!(bytes instanceof Uint8Array) || bytes.length !== 64) throw new Error("decodeCiphertext: expected 64 bytes");
  const c1 = decodePoint(bytes.subarray(0, 32));
  const c2 = decodePoint(bytes.subarray(32, 64));
  return { c1, c2 };
}

export class Reader {
  readonly bytes: Uint8Array;
  off = 0;
  constructor(bytes: Uint8Array) {
    this.bytes = bytes;
  }
  take(n: number): Uint8Array {
    if (!Number.isInteger(n) || n < 0) throw new Error("Reader.take: invalid n");
    if (this.off + n > this.bytes.length) throw new Error("Reader: out of bounds");
    const out = this.bytes.subarray(this.off, this.off + n);
    this.off += n;
    return out;
  }
  takeU8(): number {
    return this.take(1)[0]!;
  }
  takeU16LE(): number {
    return u16FromBytesLE(this.take(2));
  }
  done(): boolean {
    return this.off === this.bytes.length;
  }
}

File: apps/chain/internal/ocpshuffle/encoding.go
package ocpshuffle

import (
	"encoding/binary"
	"fmt"

	"onchainpoker/apps/chain/internal/ocpcrypto"
)

func u16ToBytesLE(x uint16) []byte {
	b := make([]byte, 2)
	binary.LittleEndian.PutUint16(b, x)
	return b
}

func u16FromBytesLE(b []byte) (uint16, error) {
	if len(b) != 2 {
		return 0, fmt.Errorf("u16FromBytesLE: expected 2 bytes")
	}
	return binary.LittleEndian.Uint16(b), nil
}

func encodePoint(p ocpcrypto.Point) []byte {
	return p.Bytes()
}

func decodePoint(b []byte) (ocpcrypto.Point, error) {
	return ocpcrypto.PointFromBytesCanonical(b)
}

func encodeScalar(s ocpcrypto.Scalar) []byte {
	return s.Bytes()
}

func decodeScalar(b []byte) (ocpcrypto.Scalar, error) {
	return ocpcrypto.ScalarFromBytesCanonical(b)
}

func encodeCiphertext(ct ocpcrypto.ElGamalCiphertext) []byte {
	return append(encodePoint(ct.C1), encodePoint(ct.C2)...)
}

func decodeCiphertext(b []byte) (ocpcrypto.ElGamalCiphertext, error) {
	if len(b) != 64 {
		return ocpcrypto.ElGamalCiphertext{}, fmt.Errorf("decodeCiphertext: expected 64 bytes")
	}
	c1, err := decodePoint(b[:32])
	if err != nil {
		return ocpcrypto.ElGamalCiphertext{}, err
	}
	c2, err := decodePoint(b[32:64])
	if err != nil {
		return ocpcrypto.ElGamalCiphertext{}, err
	}
	return ocpcrypto.ElGamalCiphertext{C1: c1, C2: c2}, nil
}

type reader struct {
	bytes []byte
	off   int
}

func newReader(b []byte) *reader {
	return &reader{bytes: b}
}

func (r *reader) take(n int) ([]byte, error) {
	if n < 0 {
		return nil, fmt.Errorf("reader.take: invalid n")
	}
	if r.off+n > len(r.bytes) {
		return nil, fmt.Errorf("reader: out of bounds")
	}
	out := r.bytes[r.off : r.off+n]
	r.off += n
	return out, nil
}

func (r *reader) takeU8() (uint8, error) {
	b, err := r.take(1)
	if err != nil {
		return 0, err
	}
	return b[0], nil
}

func (r *reader) takeU16LE() (uint16, error) {
	b, err := r.take(2)
	if err != nil {
		return 0, err
	}
	return u16FromBytesLE(b)
}

func (r *reader) done() bool {
	return r.off == len(r.bytes)
}

